sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/DoNotOverwriteCollectionElementsBase.cs

```diff
@@ -20,224 +20,100 @@
-
-            protected abstract SyntaxNode GetName(TMemberAccessSyntax memberAccess);
-
-            protected abstract SyntaxToken GetIdentifierToken(TMemberAccessSyntax memberAccess);
-
-            protected abstract TInvocationSyntax GetInvocation(SyntaxNode node);
-
-            public Action<SyntaxNodeAnalysisContext> GetAnalysisAction(DiagnosticDescriptor rule) =>
-                c =>
-                {
-                    var invocation = GetInvocation(c.Node);
-                    if (!HasNumberOfArguments(invocation, 2))
-                    {
-                        return;
-                    }
-
-                    var methodName = GetMethodName(invocation);
-                    var invokedOn = this.getInvokedOnName(invocation);
-
-                    if (string.IsNullOrWhiteSpace(methodName) ||
-                        !methodName.Equals("Add", StringComparison.OrdinalIgnoreCase) ||
-                        string.IsNullOrWhiteSpace(invokedOn) ||
-                        !IsDictionaryAdd(c.SemanticModel.GetSymbolInfo(invocation).Symbol as IMethodSymbol))
-                    {
-                        return;
-                    }
-
-                    var keyValue = GetFirstArgument(invocation).ToString();
-
-                    var previousAddInvocationOnVariable = this.getPreviousStatements(invocation)
-                         .OfType<TExpressionSyntax>()
-                         .Select(ess => GetExpression(ess))
-                         .TakeWhile(e => IsInvocationOnSameItem(e, invokedOn))
-                         .Cast<TInvocationSyntax>()
-                         .FirstOrDefault(ies => GetFirstArgument(ies).ToString() == keyValue &&
-                             IsDictionaryAdd(c.SemanticModel.GetSymbolInfo(ies).Symbol as IMethodSymbol));
-
-                    if (previousAddInvocationOnVariable != null)
-                    {
-                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, invocation.GetLocation(),
-                             additionalLocations: new[] { previousAddInvocationOnVariable.GetLocation() }));
-                    }
-                };
-
-            private bool IsInvocationOnSameItem(SyntaxNode expression, string invokedOn) =>
-                expression is TInvocationSyntax ies &&
-                HasNumberOfArguments(ies, 2) &&
-                "Add".Equals(GetMethodName(ies), StringComparison.OrdinalIgnoreCase) &&
-                invokedOn.Equals(getInvokedOnName(ies), StringComparison.OrdinalIgnoreCase);
 
-            protected string GetMethodName(TInvocationSyntax invocation) =>
-                GetExpression(invocation) is TMemberAccessSyntax memberAccess
-                    ? GetIdentifierToken(memberAccess).ValueText
-                    : null;
+            var previousSet = GetPreviousStatements(statement)
+                .TakeWhile(IsSameCollection(collectionIdentifier))
```

Comment:
All in all, this is a O(n²) algorithm, where n is the number of consecutive statements that are accessing the same collection... And each operation of this O(n²) algorithm is not trivial either...
I've already seen some code with thousands of assignments to the same array (I even wrote such code when I was very young... 😊), I wonder what would happen in this case...

I made a test by adding to the unit tests 5000 lines such as:
```c#
            result[0] = "0";
            result[1] = "1";
            result[2] = "2";
            result[3] = "3";
            result[4] = "4";
            result[5] = "5";
            result[6] = "6";
            result[7] = "7";
            result[8] = "8";
            result[9] = "9";
            result[10] = "10";
            result[11] = "11";
```
The execution of the test went from 2s to 47s.

I think we can do better if we first construct the list of all concerned statements (this should be O(n)), then check that each index is used only once (O(n), if we use a hash-table to remember already accessed indices).

Reply:
I agree, but this would mean to rewrite the rule again...

Reply:
I'll give it a shot in the afternoon if I can fix the other tickets :)